machine/wizards/machine_email.py

```diff
@@ -0,0 +1,188 @@
+        data = []
+        for group in self.machine_group_ids:
+            for machine in group.machine_instance_ids:
+                partner_ids = machine.message_get_default_recipients()[
+                    machine.id]['partner_ids']
+                data.append((
+                    0, 0,
+                    {
+                        'machine_instance_id': machine.id,
+                        'partner_ids': partner_ids,
+                    }
+                ))
+        if data:
+            self.update({'recipient_ids': data})
+
+    @api.onchange('email_type')
+    def _onchange_email_type(self):
+        if self.email_type == 'scheduled':
+            self.mail_template_id = self.env.ref(
+                'machine.mail_template_planned_machine_change').id
+        else:
+            self.mail_template_id = False
+
+    @api.multi
+    def action_create_mail_messages(self):
+        """Create mail messages with a type email for each machine.
+
+        Returns messages that were created.
+        """
+        self.check_recipients()
+        self.check_message()
+        messages = self.env['mail.message']
+        for recipient in self.recipient_ids:
+            messages |= recipient._create_mail_message()
+        return messages
+
+
+class MachineEmailRecipient(models.TransientModel):
+    """Model to hold recipient data for MachineEmail wizard."""
+
+    _name = 'machine.email.recipient'
+    _description = 'Machine Email Recipient'
+    _rec_name = 'machine_instance_id'
+
+    machine_instance_id = fields.Many2one(
+        'machine.instance',
+        "Machine Instance",
+        domain=[('parent_id', '!=', False)],
+        required=True)
+    partner_ids = fields.Many2many(
+        'res.partner',
+        string="Partners",
+        help="Partners that will receive this email")
+    machine_email_id = fields.Many2one(
+        'machine.email', "Machine Email", required=True, ondelete='cascade')
+    # Technical field used for mail template rendering.
+    change_log_id = fields.Many2one(
+        'machine.instance.change_log', "Change Log")
+
+    def _add_change_log(self):
+        """Create change log record and relate with recipient."""
+        wiz = self.machine_email_id
+        change_log = self.env['machine.instance.change_log'].create({
+            'name': wiz.sub_subject,
+            'date': wiz.date,
+            'duration': wiz.duration,
+            'user_id': wiz.user_id.id,
+            'priority': wiz.priority,
+            'machine_instance_id': self.machine_instance_id.id,
+
+        })
+        self.change_log_id = change_log.id
+
+    def _get_subject_and_body(self):
+        self.ensure_one()
+        wiz = self.machine_email_id
+        if wiz.email_type == 'general':
+            return wiz.subject, wiz.body
+        elif wiz.email_type == 'scheduled':
+            # Add change log, so we can render template from it.
+            self._add_change_log()
```

Comment:
It's just a suggestion, but wouldn't it be convenient to have a history (in machine instance or group) of emails sent?

Reply:
It will have as mail.message. Or you want some extra history somewhere?

Reply:
yes